---
title: "Escort"
author: "Xiaoru Dong"
output: 
  BiocStyle::html_document:
    toc: true
fontsize: 13pt
vignette: >
  %\VignetteIndexEntry{Escort}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r set-options, echo=FALSE, cache=FALSE}
options(width = 1000)
```


```{r, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      error = FALSE)
```


# Introduction

`Escort` is a framework for evaluating a single-cell RNA-seq datasetâ€™s suitability for trajectory inference and for quantifying trajectory properties influenced by analysis decisions. Escort is an R package designed to guide users through the trajectory inference process by offering goodness-of-fit evaluations for embeddings that represent a range of analysis decisions such as feature selection, dimension reduction, and trajectory inference method-specific hyperparameters.

# Installation

`Escort` can be installed via GitHub:

```{r, results='hide', message=FALSE, warning=FALSE, eval=FALSE}
if (!requireNamespace("devtools", quietly=TRUE))
    install.packages("devtools")
devtools::install_github("xiaorudong/Escort")
```

The following packages are required for installing Escort. If installation fails, you may need to manually install the dependencies using the function 'install.packages' for CRAN packages or 'BiocManager::install' for Bioconductor packages. The error message will indicate which packages are missing, or you can go ahead and pre-install these.

CRAN packages: alphahull, fastICA, Rtsne, umap, rstatix, DT, FNN, grDevices, Hmisc, jmuOutlier, shotGroups, parallelDist, robustbase, scales, SCORPIUS, sfsmisc shiny, shinycssloaders, shinydashboard, shinyjs, shinyWidgets, dplyr

Bioconductor packages: SingleCellExperiment, SC3, slingshot, clusterProfiler, org.Hs.eg.db, org.Mm.eg.db, edgeR, limma

Bioconductor packages can be installed using: 

```{r, results='hide', message=FALSE, warning=FALSE, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("PACKAGE NAME")
```

CRAN packages can be installed using:

```{r, results='hide', message=FALSE, warning=FALSE, eval=FALSE}
install.packages("PACKAGE NAME")
```

# Example using a simulated single-cell RNA-seq dataset

Below we show an example using a simulated single-cell RNA-seq dataset from Saelens et al., 2019, available at (https://zenodo.org/records/1443566). For this example, we pre-normalized the data and removed genes having mean expression less than three. This simulated dataset is characterized by a linear topology structure. We will also set a seed here for reproducibility.

```{r, warning=FALSE, error=FALSE, message=FALSE}
set.seed(11111)
library(Escort)
data("step0_clean_dyntoy_L3")
```

Escort first assists users by determining the presence of a trajectory signal. This step provides information regarding the appropriateness of the dataset for trajectory analysis. Often trajectory analysis is not appropriate when cells represent distinct cell types or when cells are too homogeneous.

# Step 1: Detecting Trajectory Existence 

The input for this step are the matrices of raw and normalized single-cell RNA-seq data after any quality control. The `HD_DCClusterscheck()` function is utilized to identify diverse cell types, while the `testHomogeneous()` function aids in assessing homogeneity.

### Assessing evidence of any distinct cell types:

In practice, we do not recommend setting K below, it was done here only for speed.

```{r, warning=FALSE, error=FALSE, message=FALSE}
LvsC <- HD_DCClusterscheck(normcounts=norm_counts, rawcounts=rawcounts, K=2)
LvsC$DCcheck
```

### Assessing evidence of cell homogeneity:

In practice, we recommend setting the number of iterations to at least 20. We set it to 5 here for speed.

```{r, warning=FALSE, error=FALSE, message=FALSE}
cor_test <- step1_testHomogeneous(normcounts=norm_counts, num.sim=5)
cor_test$decision
```

# Step 2: Evaluating embedding-specific characteristics

Escort's next step is to identify dataset-specific preferred embeddings for performing trajectory inference. An embedding encompasses all analysis or processing choices involved prior to trajectory inference (e.g. dimension reduction, feature selection, normalization, etc.)

For feature selection, usually the top highly variable genes are selected, although there is no pre-determined optimal number of genes to select. For dimension reduction, Escort has built-in the more well-known dimension reduction algorithms, specifically, UMAP, t-SNE, MDS, and PCA, through the `getDR_2D()` function.

Here we will evaluate an embedding consisting of selecting the top 20% of highly variable genes followed by PCA for dimension reduction. The `modelGeneVar()` function from `scran` package is used here to identify highly variable genes, although any procedure could be used.

```{r, warning=FALSE, error=FALSE, message=FALSE}
gene.var <- quick_model_gene_var(norm_counts)
genes.HVGs <- rownames(gene.var)[1:1000] # (approx. 20%)

embedding1 <- getDR_2D(norm_counts[genes.HVGs,], "PCA")
head(embedding1)
```


### Examining cell connectivity on embeddings

Escort first evaluates cell connectivity on the embedding using the `LD_DCClusterscheck()` function. Since the presence of distinct clusters w=has already been performed in the first step, then a reliable embedding should not exhibit distinct clusters. Thus, any embeddings found to be disconnected are classified immediately as Non-recommended.

```{r, warning=FALSE, error=FALSE, message=FALSE}
DRLvsC <- LD_DCClusterscheck(embedding=embedding1)
DRLvsC$DCcheck
```

### Examining preservation of cell relationships

Next, Escort assesses the effectiveness of the embedding at preserving inter-cellular relationships that were present in the high-dimensional data using the `Similaritycheck()` function. The output percentage represents the rate at which cells in the embedding successfully maintain inter-cellular relationships. A higher percentage indicates a greater preservation of relationships among cells.

```{r, warning=FALSE, error=FALSE, message=FALSE}
simi_cells <- Similaritycheck(normcounts=norm_counts, dimred=embedding1, clusters=LvsC)
simi_cells$GoodRate
```


### Examining cell density

Escort next evaluates the distribution of cells within the embedding. A tighter distribution of cells suggests an enhanced capacity to achieve higher-quality trajectory fits. In this context, we compute the cell coverage rate as a representation of cell density by applying `GOFeval()` function.

```{r, warning=FALSE, error=FALSE, message=FALSE}
gof_eval <- GOFeval(embedding1)
gof_eval$occupiedRate
```


# Step 3: Quantifying Trajectory Fitting Performance

Now that the embedding has been evaluated independently, the final step involves assessing the embedding within the framework of a specific trajectory inference method to allow for evaluating method-specific hyperparameters. In this step, Escort will evaluate the combined embedding and inference method's proportion of cells having an ambiguous projection to the preliminary trajectory.

Here, we use Slingshot for trajectory fitting:

```{r, warning=FALSE, error=FALSE, message=FALSE, fig.width=4, fig.height=4, fig.align="left"}
library(slingshot)
library(mclust)
cls <- Mclust(embedding1)$classification
ti_out <- slingshot(data=embedding1, clusterLabels=cls)
rawpse <- slingPseudotime(ti_out, na=T)
ls_fitLine <- lapply(slingCurves(ti_out), function(x) x$s[x$ord,])

library(grDevices)
library(RColorBrewer)
colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(rawpse, breaks=100)]
plot(embedding1, col = plotcol, pch=16, cex=.5)
lines(SlingshotDataSet(ti_out), lwd=2, col='black')
```

### Examining ambiguous cells

After fitting a trajectory, we use the `prepTraj()` function in Escort to generate an object that stores all relevant information about the embedding and trajectory. The fitted trajectory line is saved as segments between pairs of points within the `prepTraj()` function. Subsequently, the `UshapeDetector()` function is employed to calculate the proportion of ambiguous cells. For trajectories estimated to be smooth, such as those generated by Slingshot, it is recommended to use `outlierdetect='neutral'` in the `UshapeDetector()` function. Conversely, for convoluted trajectories like those produced by Monocle 3, it is suggested to use `outlierdetect='asymmetric'`. A smaller number of ambiguous cells indicates a more accurate estimation of pseudotime along the trajectory.

```{r, warning=FALSE, error=FALSE, message=FALSE}

fitLine <- segFormat(ls_fitLine) # needed when using Slingshot to get the preliminary trajectory

resobj <- prepTraj(embedding1, PT=rawpse, fitLine=fitLine)

ushap_eval <- UshapeDetector(resobj)
ushap_eval$Ambpct
```


# Scoring System

Finally, a comprehensive score is computed to assess the overall performance of each embedding. This score encompasses all four components evaluated in Steps 2 and 3: cell connectivity (DCcheck), preservation of cell relationships (SimiRetain), cell density (GOF), and ambiguous cells (USHAPE). Results from these steps are stored in a data frame with columns: `DCcheck`, `SimiRetain`, `GOF`, and `USHAPE`. This data frame serves as input for the `score_cal()` function. Each embedding receives a score, indicating the recommended level for constructing a trajectory. Embeddings with a score greater than zero are reported as Recommended by Escort, while those with a score less than or equal to zero are considered Non-recommended.

```{r, warning=FALSE, error=FALSE, message=FALSE}
scoredf <- data.frame(DCcheck=DRLvsC$ifConnected, SimiRetain=simi_cells$GoodRate,
                      GOF=gof_eval$occupiedRate, USHAPE=ushap_eval$Ambpct)
calcScore(scoredf)
```



# SessionInfo

```{r sessionInfo, results='markup'}
sessionInfo()
```
