---
title: "Evaluating Embeddings in Parallel with Escort"
author: "Xiaoru Dong and Rhonda Bacher"
output: 
  BiocStyle::html_document:
    toc: true
fontsize: 13pt
vignette: >
  %\VignetteIndexEntry{Escort}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r set-options, echo=FALSE, cache=FALSE}
options(width = 1000)
```


```{r, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      error = FALSE)
```

# Introduction


Here, we demonstrate how to use `Escort` to evaluate multiple embeddings and trajectories in parallel during Steps 2 and 3, assuming Step 1 was successful.


# Example using a simulated single-cell RNA-seq dataset

Below we show an example using a simulated single-cell RNA-seq dataset from Saelens et al., 2019, available at (https://zenodo.org/records/1443566). For this example, we pre-normalized the data and removed genes having mean expression less than three. This simulated dataset is characterized by a linear topology structure. 

```{r, warning=FALSE, error=FALSE, message=FALSE}
library(Escort)
set.seed(11111)
data("step0_clean_dyntoy_L3")
```

# Generating multiple embeddings

We will utilize a function to generate all embeddings, then we store them in a list which will be accessed in parallel in the following steps.

```{r, warning=FALSE, error=FALSE, message=FALSE}
gene.var <- quick_model_gene_var(norm_counts)

library(mclust)
library(slingshot)
myembeds <- function(varyg, usedim) {
  genes.HVGs <- rownames(gene.var)[1:varyg]
  embedding1 <- getDR_2D(norm_counts[genes.HVGs,], usedim)
  cls1 <- Mclust(embedding1)$classification
  ti_out1 <- slingshot(data=embedding1, clusterLabels=cls1)
  rawpse1 <- slingPseudotime(ti_out1, na=T)
  ls_fitLine1 <- lapply(slingCurves(ti_out1), function(x) x$s[x$ord,])
  
  embed_obj <- prepTraj(dimred=embedding1, PT=rawpse1, fitLine=ls_fitLine1)
  return(embed_obj)
}

pca_embeds <- lapply(c(1000, 2000, 5000), myembeds, usedim="PCA")
umap_embeds <- lapply(c(1000, 2000, 5000), myembeds, usedim="UMAP")
mds_embeds <- lapply(c(1000, 2000, 5000), myembeds, usedim="MDS")

embeddings <- c(pca_embeds, umap_embeds, mds_embeds)
```


# Parallelizing embedding evaluations

We will use the `mclapply()` function from the `parallel` package to evaluate the embeddings. Each evaluate step will be performed separately as before.


#### Examining cell connectivity on embeddings

```{r, warning=FALSE, error=FALSE, message=FALSE}
library(parallel)
n = detectCores() - 1
DRLvsCs <- mclapply(embeddings, function(x) LD_DCClusterscheck(x$Embedding), mc.cores = n)
sapply(DRLvsCs, function(x) x$DCcheck)
```

#### Examining preservation of cell relationships

```{r, warning=FALSE, error=FALSE, message=FALSE}
## This was obtained in Step 1, or can be re-run here.
LvsC <- HD_DCClusterscheck(normcounts=norm_counts, rawcounts=rawcounts, K=2)

simi_cells <- mclapply(embeddings, function(x) {
  Similaritycheck(normcounts=norm_counts, dimred=x$Embedding, clusters=LvsC)
}, mc.cores = n)
sapply(simi_cells, function(x) x$GoodRate)
```


#### Examining cell density

```{r, warning=FALSE, error=FALSE, message=FALSE}
gof_evals <- mclapply(embeddings, function(x) GOFeval(x$Embedding), mc.cores = n)
sapply(gof_evals, function(x) x$occupiedRate)
```

#### Examining ambiguous cells

```{r, warning=FALSE, error=FALSE, message=FALSE}
ushap_evals <- mclapply(embeddings, UshapeDetector, mc.cores = n)
sapply(ushap_evals, function(x) x$Ambpct)
```


# Scoring System

```{r, warning=FALSE, error=FALSE, message=FALSE}
scoredf <- data.frame(DCcheck=sapply(DRLvsCs, function(x) x$ifConnected),
                      SimiRetain=sapply(simi_cells, function(x) x$GoodRate),
                      GOF=sapply(gof_evals, function(x) x$occupiedRate), 
                      USHAPE=sapply(ushap_evals, function(x) x$Ambpct))
rownames(scoredf) <- c(paste(rep(c("PCA","UMAP","MDS"), each=3), rep(c(1000, 2000, 5000), 3)))
calcScore(scoredf)
```



# SessionInfo

```{r sessionInfo, results='markup'}
sessionInfo()
```
